python/dgl/nn/pytorch/explain/pgexplainer.py

```diff
@@ -0,0 +1,356 @@
+        ...         h = self.conv(g, h, edge_weight=edge_weight)
+        ...         if not embed:
+        ...             g.ndata['h'] = h
+        ...             hg = dgl.mean_nodes(g, 'h')
+        ...             return self.fc(hg)
+        ...         else:
+        ...             return h
+
+        >>> # Load dataset
+        >>> data = GINDataset('MUTAG', self_loop=True)
+        >>> dataloader = GraphDataLoader(data, batch_size=64, shuffle=True)
+
+        >>> # Train the model
+        >>> feat_size = data[0][0].ndata['attr'].shape[1]
+        >>> model = Model(feat_size, data.gclasses)
+        >>> criterion = nn.CrossEntropyLoss()
+        >>> optimizer = th.optim.Adam(model.parameters(), lr=1e-2)
+        >>> for bg, labels in dataloader:
+        ...     preds = model(bg, bg.ndata['attr'])
+        ...     loss = criterion(preds.squeeze(1).float(), labels.float())
+        ...     optimizer.zero_grad()
+        ...     loss.backward()
+        ...     optimizer.step()
+
+        >>> # Initialize the explainer
+        >>> explainer = PGExplainer(model, data.gclasses)
+
+        >>> # Train the explainer
+        >>> # Define explainer temperature parameter
+        >>> init_tmp, final_tmp = 5.0, 1.0
+        >>> optimizer_exp = torch.optim.Adam(explainer.parameters(), lr=0.01)
+        >>> for epoch in range(20):
+        ...     tmp = float(init_tmp * np.power(final_tmp / init_tmp, epoch / 20))
+        ...     for bg, labels in dataloader:
+        ...          loss = explainer.train_step(bg, bg.ndata['attr'], tmp)
+        ...          optimizer_exp.zero_grad()
+        ...          loss.backward()
+        ...          optimizer_exp.step()
+
+        >>> # Explain the prediction for graph 0
+        >>> graph, l = data[0]
+        >>> graph_feat = graph.ndata.pop("attr")
+        >>> probs, edge_weight = explainer.explain_graph(graph, graph_feat)
+        """
+        self.model = self.model.to(graph.device)
+        self.elayers = self.elayers.to(graph.device)
+
+        embed = self.model(graph, feat, embed=True, **kwargs)
+        embed = embed.data
+
+        edge_idx = graph.edges()
+
+        node_size = embed.shape[0]
+        col, row = edge_idx
+        col_emb = embed[col.long()]
+        row_emb = embed[row.long()]
+        emb = torch.cat([col_emb, row_emb], dim=-1)
+
+        for elayer in self.elayers:
+            emb = elayer(emb)
+        values = emb.reshape(-1)
+
+        values = self.concrete_sample(values, beta=tmp, training=training)
+        self.sparse_mask_values = values
+
+        mask_sparse = torch.sparse_coo_tensor(
+            [edge_idx[0].tolist(), edge_idx[1].tolist()],
+            values.tolist(),
+            (node_size, node_size),
+        )
+        mask_sigmoid = mask_sparse.to_dense()
+        # set the symmetric edge weights
+        sym_mask = (mask_sigmoid + mask_sigmoid.transpose(0, 1)) / 2
+        edge_mask = sym_mask[edge_idx[0].long(), edge_idx[1].long()]
```

Comment:
Consider doing

```
reverse_eids = g.edge_ids(row, col).long()
edge_mask = (values + values[reverse_eids]) / 2
```
The code may not be bug-free. Please try it first.

Reply:
Tested it, and I had to update how rol, col was extracted from the edge_idx.